// ngx-blockly/esm2020/lib/ngx-blockly.module.mjs
import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { NgxBlocklyComponent } from './ngx-blockly/ngx-blockly.component';
import * as Blockly from 'blockly/core';
import * as i0 from "@angular/core";
export class NgxBlocklyModule {
}
NgxBlocklyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: NgxBlocklyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxBlocklyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: NgxBlocklyModule, declarations: [NgxBlocklyComponent], exports: [NgxBlocklyComponent] });
NgxBlocklyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: NgxBlocklyModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: NgxBlocklyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxBlocklyComponent],
                    imports: [],
                    exports: [NgxBlocklyComponent],
                    providers: [],
                    schemas: [NO_ERRORS_SCHEMA]
                }]
        }] });
// Still needed?
// Blockly.ToolboxCategory.prototype.parseContents_ = function (categoryDef) {
//     const contents = categoryDef['contents'];
//     if (categoryDef['custom']) {
//         this.flyoutItems_ = categoryDef['custom'];
//     } else if (contents) {
//         for (let i = 0, itemDef; (itemDef = contents[i]); i++) {
//             const flyoutItem = (itemDef);
//             this.flyoutItems_.push(flyoutItem);
//         }
//     }
//     if (categoryDef['categoryclass']) {
//         this.cssConfig_.row += ' ' + categoryDef['categoryclass'];
//     }
// };
Blockly.CollapsibleToolboxCategory.prototype.parseContents_ = function (categoryDef) {
    const contents = categoryDef['contents'];
    let prevIsFlyoutItem = true;
    if (categoryDef['custom']) {
        this.flyoutItems_ = categoryDef['custom'];
    }
    else if (contents) {
        for (let i = 0, itemDef; (itemDef = contents[i]); i++) {
            // Separators can exist as either a flyout item or a toolbox item so
            // decide where it goes based on the type of the previous item.
            if (!Blockly.registry.hasItem(Blockly.registry.Type.TOOLBOX_ITEM, itemDef['kind']) ||
                (itemDef['kind'].toLowerCase() === Blockly.ToolboxSeparator.registrationName &&
                    prevIsFlyoutItem)) {
                const flyoutItem = (itemDef);
                this.flyoutItems_.push(flyoutItem);
                prevIsFlyoutItem = true;
            }
            else {
                this.createToolboxItem_(itemDef);
                prevIsFlyoutItem = false;
            }
        }
    }
    if (categoryDef['categoryclass']) {
        this.cssConfig_.row += ' ' + categoryDef['categoryclass'];
    }
};
