// ngx-blockly/esm2020/lib/ngx-blockly/models/custom-block.mjs
import { Block } from './block';
export class CustomBlock extends Block {
    constructor(type, blockMutator, ...args) {
        super(type);
        this._blockMutator = blockMutator ? blockMutator : null;
        this._args = args;
    }
    init(block) {
        this._block = block;
        this.defineBlock();
        this.block.setOnChange(function (event) {
            this.blockInstance.onChange(event);
        });
    }
    onChange(changeEvent) {
        // nothing to do
    }
    toXML() {
        return `<block type="${this.type}" disabled="${this.disabled}"></block>`;
    }
    toDartCode(block) {
        return 'Not implemented';
    }
    toJavaScriptCode(block) {
        return 'Not implemented';
    }
    toLuaCode(block) {
        return 'Not implemented';
    }
    toPHPCode(block) {
        return 'Not implemented';
    }
    toPythonCode(block) {
        return 'Not implemented';
    }
    get block() {
        return this._block;
    }
    set block(block) {
        this._block = block;
    }
    get blockMutator() {
        return this._blockMutator;
    }
    set blockMutator(mutator) {
        this._blockMutator = mutator;
    }
    get args() {
        return this._args;
    }
    set args(args) {
        this._args = args;
    }
}
