// ngx-blockly/esm2020/lib/ngx-blockly/models/xml-block.mjs
import { Block } from './block';
export class XmlBlock extends Block {
    constructor(type, shadow = false) {
        super(type);
        this._values = [];
        this._shadow = shadow;
    }
    get values() {
        return this._values;
    }
    set values(value) {
        this._values = value;
    }
    get field() {
        return this._field;
    }
    set field(value) {
        this._field = value;
    }
    get mutation() {
        return this._mutation;
    }
    set mutation(mutation) {
        this._mutation = mutation;
    }
    toXML() {
        const tagName = this._shadow ? 'shadow' : 'block';
        let xml = `<${tagName} type="${this.type}" disabled="${this.disabled}">`;
        for (const value of this.values) {
            xml += value.toXML();
        }
        xml += this.field ? this.field.toXML() : '';
        if (this._mutation) {
            xml += this._mutation.toXML();
        }
        xml += `</${tagName}>`;
        return xml;
    }
}
export class Value {
    constructor(name, block) {
        this._name = name;
        this._block = block;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get block() {
        return this._block;
    }
    set block(value) {
        this._block = value;
    }
    toXML() {
        return `<value name="${this.name}">` + this.block.toXML() + '</value>';
    }
}
export class Field {
    constructor(name, value) {
        this._name = name;
        this._value = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get value() {
        return this._value;
    }
    set value(value) {
        this._value = value;
    }
    toXML() {
        return `<field name="${this.name}">` + this.value + '</field>';
    }
}
export class Mutation {
    constructor(name, value) {
        this._name = name;
        this._value = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get value() {
        return this._value;
    }
    set value(value) {
        this._value = value;
    }
    toXML() {
        return '<mutation ' + this.name + ' = "' + this.value + '"></mutation>';
    }
}
